contracts/src/set_verifier.rs

```diff
@@ -116,4 +133,110 @@ where
 
         Ok((image_id, image_url))
     }
+
+    /// Returns the seal if of the given verified root.
+    pub async fn get_verified_root_seal(&self, root: B256) -> Result<Bytes> {
+        self.query_verified_root_event(root, None, None).await
+    }
+
+    async fn get_latest_block(&self) -> Result<u64> {
+        self.instance
+            .provider()
+            .get_block_number()
+            .await
+            .context("Failed to get latest block number")
+    }
+
+    /// Query the VerifiedRoot event based on the root and block options.
+    /// For each iteration, we query a range of blocks.
+    /// If the event is not found, we move the range down and repeat until we find the event.
+    /// If the event is not found after the configured max iterations, we return an error.
+    /// The default range is set to 100 blocks for each iteration, and the default maximum number of
+    /// iterations is 100. This means that the search will cover a maximum of 10,000 blocks.
+    /// Optionally, you can specify a lower and upper bound to limit the search range.
+    async fn query_verified_root_event(
+        &self,
+        root: B256,
+        lower_bound: Option<u64>,
+        upper_bound: Option<u64>,
+    ) -> Result<Bytes> {
+        let mut upper_block = upper_bound.unwrap_or(self.get_latest_block().await?);
```

Comment:
`unwrap_or` is evaluated eagerly, meaning this will make an API call every time, even if an upper-bound is provided. I wish there was a native async version of `unwrap_or_else`, but there is not.
```suggestion
        let mut upper_block = if let Some(upper_bound) = upper_bound else { self.get_latest_block().await? };
```